### PR TITLE
fix(markdown): avoid unnecessary destruction and recreation of custom elements when value is updated

### PR DESCRIPTION
**Pull request by adrianschmidt**
## Summary
- Replaces `innerHTML` with [morphdom](https://github.com/patrick-steele-idem/morphdom) in `limel-markdown` to diff the DOM instead of destroying and recreating it on every value update
- Existing custom elements (e.g. `limebb-object-chip`) are preserved in the DOM across updates, preventing lifecycle churn and visual flickering
- Particularly impactful during streaming updates (AI chat), where the value changes every ~50ms

## Background
When `limel-markdown` is used in the AI chat to render streaming responses, every value update destroyed and recreated the entire DOM tree. Custom elements like `limebb-object-chip` would have their `disconnectedCallback` and `connectedCallback` fired on every update, triggering data refetching and causing severe flickering.

fix: Lundalogik/crm-insights-and-intelligence#237

## Test plan
- [ ] Verify all markdown component examples render correctly in the docs site
- [ ] Verify custom component examples (`limel-example-markdown-custom-component`, `limel-example-markdown-custom-component-with-json-props`) still hydrate properly
- [ ] Verify AI chat streaming with `limebb-object-chip` no longer flickers
- [ ] CI tests pass

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Markdown rendering now uses intelligent DOM updates that preserve existing custom elements and node references during content changes, improving stability for dynamic markdown content.

* **Tests**
  * Added end-to-end test coverage for markdown component updates and DOM morphing behavior, including scenarios with custom elements and attribute changes.

* **Chores**
  * Added morphdom package to development dependencies.

### CODE REVIEW

**Comment by civing** [comment]
## PR Review (via Claude Code)

### Why morphdom? (the external dependency question)

morphdom is justified here. DOM tree diffing — matching old and new nodes, diffing attributes, handling reordering — is a well-understood but non-trivial algorithm. A minimal inline implementation would be 150+ lines of subtle code with its own bug surface. morphdom has been solving exactly this problem for 8+ years (~5K GitHub stars, zero dependencies, 4KB minified). That 4KB is a rounding error in a library already bundling lodash, moment, material-components-web, and prosemirror. The thin wrapper in `morph-dom.ts` provides a clean seam if you ever need to swap it out.

---

### Must Fix

- **Remove AI working documents from the repo.** The files under `docs/superpowers/plans/` and `docs/superpowers/specs/` are transient agent planning artifacts containing hardcoded local paths (`/Users/adrian.schmidt/src/lime-elements`). They add noise and will rot. These should not be committed.

- **Test custom elements with complex JSON attributes.** The real use case is `<limebb-object-chip link='{"href":"...","text":"..."}'>` but the only custom element test uses `data-x="1"`. A morphdom/HTML-parser edge case with embedded JSON would go undetected.

- **Test attribute mutation on a preserved custom element.** What happens when a custom element's attribute *changes* between morphs? During streaming, partial JSON attributes get completed. No test covers whether morphdom mutates in-place vs. destroys/recreates.

### Should Fix

- **Double hydration on preserved elements.** `hydrateCustomElements` runs unconditionally after every morph. When morphdom preserves an existing custom element (the whole point), hydration re-parses JSON attributes and re-assigns JS properties. This could trigger unnecessary re-renders or overwrite runtime state. Consider marking already-hydrated elements or diffing before re-hydrating.

- **No error handling / fallback.** If `html` is malformed, morphdom will silently produce unexpected DOM. A try/catch falling back to `innerHTML` would ensure the component at least still renders (even if it flickers).

- **Missing tests for:** node removal (3 children → 2), nested custom elements (inside `<p>` or `<li>`), multiple same-tag custom elements (two `<limebb-object-chip>` with different attributes), insertion before a custom element (the exact streaming scenario where new content appears above existing chips), and rapid successive calls (~50ms updates).

### Minor

- **Event listener accumulation.** With innerHTML, listeners were always cleared. With morphdom, preserved nodes retain listeners. Consumer code that re-attaches listeners after every `value` change might now accumulate duplicates. Edge case since shadow DOM limits external access.

- **Text-only content untested.** `morphChildren(container, 'just plain text')` — markdown that is a single line with no block markup.

- **Wrapper `<div>` assumption.** The `<div>${html}</div>` trick works because the container is always a plain `<div>`. Worth a comment noting this constraint for future maintainers.

---

**Overall:** The change is well-scoped — a one-line swap in `markdown.tsx` backed by a clean wrapper module. morphdom is the right tool for the job. The main concerns are the missing test scenarios (especially around the real-world JSON-attribute custom elements) and the unconditional re-hydration partially defeating the benefit of DOM preservation. The plan/spec docs should be dropped from the PR.

🤖 *This review was generated with [Claude Code](https://claude.ai/code)*

**Comment by adrianschmidt** [comment]
Thanks for the review! Here's how we addressed each point:

**Must Fix:**
- ✅ **Remove AI working documents** — Done (removed in a prior push).
- ⚡️ **Test custom elements with complex JSON attributes / attribute mutation** — We added these tests but discovered that `morphdom` hangs when imported in vitest's stencil environment (mock-doc), blocking the entire test suite. Removed the spec file for now (577970e). There is currently no automated test coverage for `morphChildren` — this needs to be addressed separately once the vitest/morphdom compatibility issue is resolved.

**Should Fix:**
- ⚡️ **Error handling / fallback** — Added. `morphChildren` now catches morphdom errors and falls back to `innerHTML`, so content is always visible even if DOM diffing fails (6bb46e1).
- ⏭️ **Double hydration on preserved elements** — Not addressing. `hydrateCustomElements` reads attributes and sets JS properties. Stencil diffs prop values internally, so re-setting the same value is a no-op. Adding hydration tracking would add complexity for no real benefit.
- ⏭️ **Missing test scenarios** — Blocked by the same vitest/morphdom hang issue described above.

**Minor:**
- ⚡️ **Wrapper `<div>` assumption** — Clarified the comment: the `<div>` wrapper is only needed to satisfy morphdom's single-root requirement. With `childrenOnly: true`, morphdom skips the root entirely, so the container can be any element type (577970e).
- ⏭️ **Event listener accumulation** — Not a concern. Custom elements in markdown are Stencil web components with shadow DOM that manage their own listeners.
- ⏭️ **Text-only content** — morphdom handles text nodes the same as element nodes, and the innerHTML fallback would catch any edge case regardless.

File: src/components/markdown/morph-dom.spec.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        const widget = container.querySelector('my-widget');
+
+        morphChildren(
+            container,
+            '<p>Before updated</p><my-widget data-x="1"></my-widget><p>After</p><p>New</p>',
+        );
+
+        const updatedWidget = container.querySelector('my-widget');
+        expect(updatedWidget).toBe(widget);
+    });
```
[the remark this answers is not in the document]
Comment (adrianschmidt) [reply src/components/markdown/morph-dom.spec.ts]: This file has been removed, because it caused the tests to hang, so blocked the CI entirely.

**Comment by civing** [comment]
## Round 2 Review (via Claude Code)

I've tested this manually — no flicker, works great. **This is ready to merge.** ✅

Below are findings from a deep review for completeness. Nothing is blocking.

---

### Should Fix (follow-up, not blocking)

- **Hydration creates fresh object references on preserved elements.** `hydrateCustomElements` runs `JSON.parse` on every whitelisted element's attributes on every value change, producing new object references. Even though morphdom preserved the DOM node, Stencil sees a "new" object (reference comparison) and re-renders the component. This partially defeats the purpose of morphdom for custom elements with object/array props. Consider tracking which elements morphdom actually modified (via `onElUpdated`) or comparing parsed values before assigning. Good candidate for a follow-up PR.

- **Guard against null/undefined `html` in `morphChildren`.** If `html` is `undefined`/`null`, the template literal produces `<div>undefined</div>`. Not reachable today (caller defaults `value` to `''`), but a `html ?? ''` guard would be cheap insurance since `morphChildren` is an exported function.

### Minor

- **Partial mutation before throw.** If morphdom throws mid-operation, the DOM may be partially morphed. The `catch` → `innerHTML` fallback correctly resets to the desired state. Sound recovery strategy.

- **Image observer disconnect/reconnect is slightly wasteful.** Already-loaded images get re-observed after morph. Harmless, and the current cleanup/setup cycle is correct.

- **No new XSS surface.** morphdom parses HTML via native `innerHTML` on a detached element — same path as before. Upstream sanitizer remains the primary defense.

### Dependency risk: Low

- Zero transitive dependencies, MIT license, ~598K weekly npm downloads
- Actively maintained (4 releases in the past year, latest Jan 2026)
- No known CVEs, ~4KB gzipped (<1% of bundle)
- `devDependencies` placement is correct for Stencil's Rollup bundling

---

🤖 *Findings generated with [Claude Code](https://claude.ai/code). Manual testing done by @civing.*

File: src/components/markdown/morph-dom.spec.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        morphChildren(container, '');
+        expect(container.innerHTML).toBe('');
+    });
+
+    it('does not replace the container element itself', () => {
+        const originalContainer = container;
+        container.innerHTML = '<p>Old</p>';
+        morphChildren(container, '<p>New</p>');
+        expect(container).toBe(originalContainer);
+    });
```
[the remark this answers is not in the document]
Comment (adrianschmidt) [reply src/components/markdown/morph-dom.spec.ts]: This file has been removed, because it caused the tests to hang, so blocked the CI entirely.

**Comment by adrianschmidt** [comment]
@coderabbitai review

File: src/components/markdown/morph-dom.e2e.ts [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    });
+
+    it('falls back to innerHTML if morphdom throws', () => {
+        // Corrupt the container in a way that might cause morphdom issues,
+        // then verify that content is still rendered via the fallback.
+        // We test this indirectly: if morphChildren never throws to the
+        // caller, the fallback is working.
+        morphChildren(container, '<p>Safe content</p>');
+        expect(container.innerHTML).toBe('<p>Safe content</p>');
+    });
```
[the remark this answers is not in the document]
Comment (adrianschmidt) [reply src/components/markdown/morph-dom.e2e.ts]: @coderabbitai: Has this been satisfactorily addressed?

**Review by civing — APPROVED** [review]
LGTM!
